pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go

```diff
@@ -28,4 +36,76 @@ type kubeEndpointsController interface {
 	InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
 		labelsList labels.Collection) ([]*model.ServiceInstance, error)
 	GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance
+	updateEDS(obj interface{}, event model.Event)
+}
+
+type kubeEndpoints struct {
+	c        *Controller
+	informer cache.SharedIndexInformer
+}
+
+// GetProxyServiceInstances returns service instances of the given proxy.
+func (e *kubeEndpoints) GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance {
+	var otherNamespaceEndpoints []*model.ServiceInstance
+	sameNamespaceEndpoints := make([]*model.ServiceInstance, 0)
+
+	for _, item := range e.informer.GetStore().List() {
+		var proxyServiceInstances []*model.ServiceInstance
+		namespace := ""
+		switch item := item.(type) {
+		case *v1.Endpoints:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpoint(c, *item, proxy)
+		case *discoveryv1alpha1.EndpointSlice:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpointSlice(c, item, proxy)
+		}
+		if namespace == proxyNamespace {
+			sameNamespaceEndpoints = append(sameNamespaceEndpoints, proxyServiceInstances...)
+		} else {
+			if otherNamespaceEndpoints == nil {
+				otherNamespaceEndpoints = make([]*model.ServiceInstance, 0)
```

Comment:
not needed, you can do `append(nil, ...)`. L50 is technically not needed either.